### PR TITLE
Add sign_raw()

### PR DESCRIPTION
**Pull request by MHHukiewitz**
`BaseAccount.sign_raw()` accepts raw bytes and outputs a signature as a string in the preferred encoding.

This allows signing of arbitrary data besides Aleph messages, making it a useful tool in dApp development.

### CODE REVIEW

**Comment by MHHukiewitz** [comment]
Relies on #49

**Comment by hoh** [comment]
The diff looks good to me. Can you edit the commits in order to have:

1. [fix formatting, flake8 bugs and imports](https://github.com/aleph-im/aleph-sdk-python/pull/50/commits/1329ce42139052b0908607d58ec3254275f15b4e)
2. The actual change of this PR by squashing the rest of the commits ?

**Comment by hoh** [comment]
The diff looks good to me. Do we have any test on the new `sign_raw` method ?

**Comment by hoh** [comment]
> Relies on #49

Squashed and pushed on master. Can you rebase on it ?

**Comment by hoh** [comment]
As discussed, `sign_raw` should return Bytes. The rest looks good to me.

**Comment by MHHukiewitz** [comment]
After suggestion by @hoh I changed the signature of `sign_raw(buffer: bytes) -> str:` to `sign_raw(buffer: bytes) -> bytes:`.

As this involves a bit more invasive refactoring, I decided that it would only make sense to add more unit tests, especially to the non-existent Cosmos and Substrate (Polkadot) tests.

File: src/aleph/sdk/chains/cosmos.py [hunk 1]
```diff
@@ -74,6 +74,7 @@ async def sign_raw(self, buffer: bytes) -> bytes:
         return base64.b64encode(signature_compact)
 
     def get_address(self) -> str:
+        # WARNING: Fails with OpenSSL >= 3.2.0 due to deprecation of ripemd160
```
Comment (hoh) [inline src/aleph/sdk/chains/cosmos.py:77]: Is that a big issue in the future ? If so, how can it be solved ?

Comment (MHHukiewitz) [reply src/aleph/sdk/chains/cosmos.py:77]: Following the discussion in Issue https://github.com/aleph-im/aleph-sdk-python/issues/23 and seeing that the cosmospy forks are not being maintained anymore, I'd suggest forking our own and adjusting it with the non-constant-time implementation of ripemd160 given in the Bitcoin repo.

Comment (hoh) [reply src/aleph/sdk/chains/cosmos.py:77]: This comment does not add any issue to the content of this branch, closing the comment.

File: src/aleph/sdk/chains/substrate.py [hunk 2]
```diff
@@ -1,47 +1,73 @@
...
+        if (
+            default_mnemonic_path.is_symlink()
+            and not default_mnemonic_path.resolve().exists()
+        ):
+            default_mnemonic_path.unlink()
+            logger.warning("The symlink to the mnemonic is broken")
+
+        # Create a symlink to use this mnemonic by default
+        if not default_mnemonic_path.exists():
+            default_mnemonic_path.symlink_to(path)
```
Comment (hoh) [inline src/aleph/sdk/chains/substrate.py:70]: This logic is not related to adding `sign_raw` and requires further discussion, especially the creation of a harcoded `default.mnemonic` file . Can you move it into an other Pull Request ?

File: src/aleph/sdk/chains/substrate.py [hunk 3]
```diff
@@ -50,6 +76,31 @@ def verify_signature(
...
-    raise NotImplementedError("Not implemented yet")
+) -> None:
+    if isinstance(signature, str):
+        if signature.startswith("0x"):
+            signature = signature[2:]
+        signature = bytes.fromhex(signature)
+    if isinstance(public_key, str):
+        if public_key.startswith("0x"):
+            public_key = public_key[2:]
+        public_key = bytes.fromhex(public_key)
```
Comment (hoh) [inline src/aleph/sdk/chains/substrate.py:87]: Would it make sense to deduplicate this ?

File: src/aleph/sdk/chains/substrate.py [hunk 4]
```diff
@@ -50,6 +76,31 @@ def verify_signature(
...
+    if isinstance(message, str):
+        message = message.encode()
+
+    try:
+        verified = verify(signature, message, public_key)
+        if not verified:
+            # Another attempt with the data wrapped, as discussed in https://github.com/polkadot-js/extension/pull/743
+            verified = verify(signature, b"<Bytes>" + message + b"</Bytes>", public_key)
+        if not verified:
+            raise BadSignatureError
```
Comment (hoh) [inline src/aleph/sdk/chains/substrate.py:97]: Could this be written in a smarter way using `or` ?

```suggestion
        # Two different attempts at verifying the data wrapped, as discussed in https://github.com/polkadot-js/extension/pull/743
        if not (
            verify(signature, message, public_key) 
            or verify(signature, b"<Bytes>" + message + b"</Bytes>", public_key)
        ):
            raise BadSignatureError
```

File: src/aleph/sdk/chains/substrate.py (hunk 2 above)
Comment (MHHukiewitz) [reply src/aleph/sdk/chains/substrate.py:70]: I added it, as it would make testing difficult. The default behavior would potentially lead to some problems, given that there might have been a raw secret key instead of a mnemonic inside.

All chains share the same private_key_file which I find a bit troubling. This should be subject to another PR, but regarding this change, I don't see how to test it otherwise.

File: src/aleph/sdk/chains/nuls1.py [hunk 5]
```diff
@@ -315,6 +315,10 @@ async def sign_message(self, message):
         message["signature"] = sig.serialize().hex()
```
[the remark this answers is not in the document]
Comment (MHHukiewitz) [reply src/aleph/sdk/chains/nuls1.py:315]: Because those two use different implementations for signing `NulsSignature.sign_message` and `NulsSignature.sign_data`

File: src/aleph/sdk/chains/cosmos.py (hunk 1 above)
Comment (odesenfans) [reply src/aleph/sdk/chains/cosmos.py:77]: For reference, I had to make a workaround for this issue for pyaleph. The CI reactivates ripemd within OpenSSL by overriding `/etc/ssl/openssl.cnf`.

**Review by odesenfans — APPROVED** [review]
